### ISSUE TITLE
Unable to use SecurityContextPersistenceFilter outside security filter chain to manage security context lifecycle

### ISSUE THREAD

**Issue by KatarzynaRadkowska** [spring-projects/spring-security#6222]
### Summary

I would like to be able to read from Security Context after security chain is done (e.g. in filters). 

Registering SecurityContextPersistenceFilter outside security chain is possible, so I thought that reading from security context, before it's cleared inside SecurityContextPersistenceFilter finally block, should be possible. 
Unfortunately security context is cleared by default in FilterChainProxy, which means that I will not be able to read from it, even before SecurityContextPersistenceFilter actually calls `SecurityContextHolder.clearContext()`.

### Actual Behavior

Security Context is cleared always in FilterChainProxy finally block.

### Expected Behavior

As SecurityContextPersistenceFilter seems to be designed specifically also for cleaning security context, this should be done only by this filter. 

### Version

spring-boot-starter-security 2.2.1.RELEASE

### Sample
  ```
// ... configuration class

  private static final int FIRST = Integer.MIN_VALUE;
  private static final int SECOND = Integer.MIN_VALUE + 1;

@Bean
  FilterRegistrationBean registerSecurityPersisterFilter() {
    FilterRegistrationBean registration = new FilterRegistrationBean<>(new SecurityContextPersistenceFilter());
    registration.setOrder(FIRST);
    return registration;
  }

  @Bean
  FilterRegistrationBean registerReadFromSecurityContext() {
    FilterRegistrationBean registration = new FilterRegistrationBean<>(new ReadFromSecurityContextFilter());
    registration.setOrder(SECOND);
    return registration;
  }
```

```
class ReadFromSecurityContextFilter extends OncePerRequestFilter {

@Override
  protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
      throws IOException {
    try {
      filterChain.doFilter(request, response);
    } finally {
      SecurityContextHolder.getContext(); // this one is already empty, even after successful auth
    }
  }
}
```

**Comment by rwinch** [comment]
Thanks for the report @KatarzynaRadkowska! Can you please put together a complete but minimal sample of this failing? This will help us to be able to prioritize the issue.

**Comment by KatarzynaRadkowska** [comment]
Hi
Working gradle example inside this zip file
[example.zip](https://github.com/spring-projects/spring-security/files/2648811/example.zip)

As you may see I try read from security context in `ReadFromSecurityContextFilter` after request processing, but still before finally block inside SecurityContextPersistenceFilter.

**Comment by rwinch** [comment]
Thanks for the sample. The FilterChainProxy clears the `SecurityContextHolder` because users can disable the `SecurityContextPersistenceFilter`. In fact, users often created memory leaks for themselves in the past by using security none (XML) or `WebSecurity.ignoring` and then trying to access the `SecurityContextHolder`. SO this is very much by design.

What are you doing that you need to access the `SecurityContext` before the `FilterChainProxy`? Could you insert any logic into the `FilterChainProxy` instead?

**Comment by KatarzynaRadkowska** [comment]
Thank you for full explanation.

Actually I'm trying to read from it after request is fully processed, so after FilterChainProxy.

The basic idea is that I'd like to log some metrics after operation is executed, but in those metrics I'd like to provide information **who** performed **which** operation, and **how long** it took. This information **who**, I will have only after successful authentication (here in the sample there was basic auth, but actually I use token, which is checked  in other system which provides me full session info). 
If I create and log those metrics after authentication then I will lose actual time of session validation and all those requests where valid token wasn't provided, won't be logged at all (in this case I can log "UNKNOWN" as **who**). 

So maybe another question: is it possible at all to have this information in the context after security filter chain? I can use some workaround and add this information to request context or to http request attributes (as it is done to mark that ...OnePerRequestFilter... has already done its job), but I consider this solution as not elegant :)

**Comment by rwinch** [comment]
Thanks for the follow up @KatarzynaRadkowska!

I don't think placing it in Spring Security's `FilterChainProxy` is any less elegant than using the `SecurityContextPersistenceFilter` outside of `FilterChainProxy`.

You can do something like this:

```java
@Override
public void configure(H http) throws Exception {
    http
       // ...
      .addFilterBefore(myFilter, ChannelProcessingFilter.class);
}
```

You would also ensure to create a `FilterRegistrationBean` that disables registering your `Filter` if it is exposed as a Bean.

An alternative is to not rely on the `SecurityContextHolder, but to instead look up the value using the `SecurityContextRepository` directly.

I'm closing this issue as this is expected behavior and there are ways to achieve the goal you want. Thanks again for submitting this ticket to help make Spring Security better!
